src/main/java/com/duke/bot/Duke.java

```diff
@@ -0,0 +1,163 @@
+package com.duke.bot;
+
+import java.lang.String;
+import java.time.LocalDate;
+
+/**
+ * Represents the Duke bot that manages the tasks of the users.
+ */
+public class Duke {
+    private TaskList tasks;
+    private DukeUi ui;
+    private Storage storage;
+
+    /**
+     * Creates a Duke Bot object.
+     */
+    public Duke() {
+        this.tasks = TaskList.createTaskList();
+        storage = Storage.createSrorageFile();
+        this.ui = new DukeUi(System.in, System.out);
+    }
+
+    /**
+     * Takes the user's input and decides the action to be taken by Duke bot.
+     *
+     * @param ui Tha user interface used by the Duke Bot to interact with the user.
+     */
+    public Duke(DukeUi ui) {
+        this.tasks = TaskList.createTaskList();
+        storage = Storage.createSrorageFile();
+        this.ui = ui;
+    }
+
+    private void echo() {
+        String userCommand = ui.getNext();
+
+        try {
+            switch (userCommand) {
+            case "list":
+                ui.print(tasks.printList());
+                ui.printEmptyLine();
+                ui.setToken("\\p{javaWhitespace}+");
+                storage.saveToFile(tasks.printList());
+                echo();
+                break;
+
+            case "done":
+                int targetIdx = ui.getNextInt() - 1;
+                this.tasks.getTask(targetIdx).markDone();
```

Comment:
Crashes if targetInx > length of tasklist || targetInx is NaN
Please prevent code from crashing at all.
Problem also present in other places.